src/prefect/results.py

```diff
@@ -73,18 +72,66 @@ def DEFAULT_STORAGE_KEY_FN():
 
 
 @sync_compatible
-async def get_default_result_storage() -> ResultStorage:
+async def get_default_result_storage() -> WritableFileSystem:
     """
     Generate a default file system for result storage.
     """
     default_block = PREFECT_DEFAULT_RESULT_STORAGE_BLOCK.value()
 
     if default_block is not None:
-        return await Block.load(default_block)
+        return await resolve_result_storage(default_block)
 
     # otherwise, use the local file system
     basepath = PREFECT_LOCAL_STORAGE_PATH.value()
-    return LocalFileSystem(basepath=basepath)
+    return LocalFileSystem(basepath=str(basepath))
+
+
+@sync_compatible
+async def resolve_result_storage(
+    result_storage: ResultStorage,
+) -> WritableFileSystem:
+    """
+    Resolve one of the valid `ResultStorage` input types into a saved block
+    document id and an instance of the block.
+    """
+    from prefect.client.orchestration import get_client
+
+    client = get_client()
+    if isinstance(result_storage, Block):
+        storage_block = result_storage
+
+        if storage_block._block_document_id is not None:
+            # Avoid saving the block if it already has an identifier assigned
+            storage_block_id = storage_block._block_document_id
+        else:
+            storage_block_id = None
+    elif isinstance(result_storage, str):
+        storage_block = await Block.load(result_storage, client=client)
```

Comment:
For the sake of not having API calls in `__init__` methods, is there any way we could defer this to when it's needed within the factory? or is that going to be a PITA?

Reply:
This is used in `update_for_flow` and `update_for_task` and we don't retrieve the default block unless `created_result` is called an no storage block has been provided, so resolution is somewhat delayed. Delaying it more would make the typing on `ResultFactory` a little weird because it would also need to be able to accept a string, but it would doable.

Reply:
That's something we could refactor later